hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -1985,4 +1991,29 @@ public static Optional<BucketCache> getBucketCacheFromCacheConfig(CacheConfig ca
     return Optional.empty();
   }
 
+  @Override
+  public Optional<List<HStoreFile>> getFilesWithStaleBlocks() {
+    return Optional.of(filesWithStaleBlocks);
+  }
+
+  public static void clearFilesWithStaleBlocks() {
+    filesWithStaleBlocks.clear();
+  }
+
+  public void setFilesWithStaleBlocks(HRegion hRegion) {
+    for (HStore hs : hRegion.getStores()) {
+      filesWithStaleBlocks.addAll(hs.getStorefiles());
```

Comment:
Shouldn't we check first which files from the region's store have blocks in the cache? If the total cache size is smaller than the store file size, we might not be able to cache all files.

Also, I wonder what's the heap usage impact of keeping such extra list. Could you check that with a heap dump?